### PR TITLE
[Bugfix] Fix EVS implementation for Qwen3 VL

### LINKED ISSUE #30847: [Bug]: Qwen 3VL via Efficient Video Sampling (EVS) to trim video embeddings and found that the number of tokens after timestamp in the Prompt was not aligned with the actual number of tokens after pruning?
### Your current environment

<details>
vllm serve Qwen3-VL-8B --video-pruning-rate=0.75 

messages=[
            {
                "role": "user",
                "content": [
                    # {"type": "text", "text": "What's in this video?"},
                    {"type": "text", "text": "这个视频和图片分别描述的是什么内容?"},
                    {
                        "type": "video_url",
                        "video_url": {
                            "url": "file:///codes/data/video/Tom_Jerry.mp4",
                            "fps": 1,
                        },
                    }
                ],
            }
        ],
<summary>The output of <code>python collect_env.py</code></summary>

```text
The get-video_deplacement_qwen3vl method in the qwen3-vl.py file
Firstly: Calculate the number of frames per frame
Secondly, add the specific timestamp of<{cur_time:. 1f} s>to the Prompt and add the calculated number of tokens after the timestamp.
At this point, the number of tokens per frame is calculated based on the clipping rate, so except for the first frame, the number of tokens after each frame remains unchanged (EVS is not used to calculate the actual tokens here).

The EVS algorithm calculates that the number of tokens reserved for each frame is different. It will cause the number of tokens after timestamp to be inconsistent with the actual number of tokens after clipping
```

</details>
、

### 🐛 Describe the bug

1、get_video_replacement_qwen3vl 
frames_idx_token=[165, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33, 33]

Image
2、compute_retention_mask

Image

Image

3、embed_input_ids
Image
input_ids:
Image

From the above 1 and 3, it can be seen that the data in frames_idx_token is the same as that in embed_input_ids,
The first frame contains 165 tokens, while the rest contain 33 tokens
151656 is the ID of the video token. The number of 151656 is the number of video tokens. The sum of the number of video tokens in all frames is the same as the sum of frames_idx_token.
Regarding the second item: compute_contention_mask EVS cropped mask, it was found that the number of tokens in the first frame was 165, while the number of tokens in other frames was different,
Based on the above 1, 2, and 3, it can be concluded that the current implementation of EVS pruning algorithm has problems
That is, the number of tokens after timestamp in the Prompt does not match the actual number of tokens that should be retained after EVS pruning.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by 2ez4bz**
## Purpose

This should fix #30847.

* Why?

Qwen3 VL requires a different EVS implementation approach compared to
Qwen2.5 VL due to its interleaved token structure.

In Qwen3 VL, video frames are separated by timestamp text tokens (e.g.,
"<0.5 seconds>") and vision marker tokens, rather than being contiguous
embeddings. This requires embedding text tokens alongside video tokens
and tracking their positions separately, which the existing EVS
implementation did not support.

* What?

This commit fixes this by heavily refactoring Qwen3 VL to generate
video embeddings separately from text embeddings (timestamps,
vision markers).

## Test Plan

Added unit tests for the mrope recomputation (see `tests/model_executor/test_qwen3_vl_mrope.py`) and some internal integration tests. Happy to run more integration tests (within reason).

## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### Accuracy

I ran a subset of the VideoMME benchmark against `nvidia/Cosmos-Reason2-8B` which is a finetuned `Qwen/Qwen3-VL-8B-Instruct`. Example command:

```
lmms-eval \
--model server.oai_server \
--tasks  videomme \
--batch_size 10 \
--log_samples
--log_samples_suffix _videomme \
--output_path <OUT_DIR> \
--verbosity INFO \
--model_args "server_kwargs.url=http://localhost:8123/v1/chat/completions,server_kwargs.timeout=180,server_kwargs.async_limit=10,server_kwargs.connection_retries=20,server_kwargs.max_retries=20,model_id=nvidia/Cosmos-Reason2-8B,continual_mode=true,response_persistent_folder=./pre/videomme_baseline"  --limit
  500 --gen_kwargs "temperature=0.0"
```

Results:

| Pruning ratio | videomme_perception_score | Pre or post change |
| ------------- | ------------------------- | ------------------ |
| 0.0           | 77.0                      | Pre                |
| 0.0           | 76.8                      | Post               |
| 0.25          | 73.0                      | Pre                |
| 0.25          | 75.8                      | Post               |
| 0.5           | 73.4                      | Pre                |
| 0.5           | 75.4                      | Post               |
| 0.75          | 70.0                      | Pre                |
| 0.75          | 72.0                      | Post               |

### Performance metrics

Using the same model, I ran some performance benchmarks using the following commands:

```sh
vllm serve \
--model nvidia/Cosmos-Reason2-8B \
--mm-processor-cache-gb 0 \
--no-enable-prefix-caching \
--media-io-kwargs '{"video":{"num_frames": -1}}' \
--mm-processor-kwargs '{"do_sample_frames": false}' \
--port 8123
```

and

```
MODEL=nvidia/Cosmos-Reason2-8B

CONC=1
NUM_PROMPTS=$((CONC * 20))

aiperf profile \
--model $MODEL \
--concurrency ${CONC} \
--endpoint v1/chat/completions \
--endpoint-type chat \
--url http://localhost:8123 \
--request-count ${NUM_PROMPTS} \
--synthetic-input-tokens-mean 30 \
--synthetic-input-tokens-stddev 0 \
--output-tokens-mean 128 \
--streaming \
--extra-inputs min_tokens:128 \
--extra-inputs ignore_eos:true \
--artifact-dir <OUTPUT_DIR> \
--video-width 512 \
--video-height 512 \
--video-fps 2 \
--video-duration 35.0 \
--video-synth-type moving_shapes \
--video-codec libx264 \
--video-format mp4 \
--num-warmup-requests 1 \
--use-server-token-count
```

This was run on a single H100 80GB SXM card.

| Pruning ratio | TTFT (ms) concurrency = 1 | Output TPS concurrency = 32|
| ------------- | ------------------------- | -------------------------- |
| 0.0           | 1207.89                   | 153.22                     |
| 0.25          | 1156.44                   | 168.94                     |
| 0.5           | 1088.46                   | 174.68                     |
| 0.75          | 1030.55                   | 177.68                     |

### CODE REVIEW

**Comment by 2ez4bz** [comment]
cc @BloodAxe :)

File: vllm/model_executor/models/qwen2_vl.py [hunk 1]
```diff
@@ -753,6 +753,8 @@ def _qwen2vl_field_config(hf_inputs: Mapping[str, torch.Tensor]):
                 "video", video_embed_grid_sizes
             ),
             video_grid_thw=MultiModalFieldConfig.batched("video", keep_on_cpu=True),
+            # TODO: double-check `keep_on_cpu`.
```
Comment (2ez4bz) [inline vllm/model_executor/models/qwen2_vl.py]: Remove.

File: vllm/model_executor/models/interfaces.py [hunk 2]
```diff
@@ -57,6 +58,26 @@
 """
 
 
+MultiModalPositions: TypeAlias = list[Tensor] | Tensor | tuple[Tensor, ...]
+
+MultiModalEmbeddingsPostProc: TypeAlias = Callable[
```
Comment (2ez4bz) [inline vllm/model_executor/models/interfaces.py]: This is an unfortunate workaround for 2 issues:

1. The first one arises from additional estimated GPU memory usage. See comments in `v1/worker/gpu_model_runner.py` for more details.
2. The second one, which I believe might be an issue for Qwen2.5 VL as well, stems from lossy dtype conversion when "smuggling" mrope position IDs in the embeddings tensor as extra channels. See comments in `vllm/model_executor/models/qwen3_vl.py` for more details.

File: vllm/model_executor/models/qwen3_vl.py [hunk 3]
```diff
@@ -1094,6 +1147,69 @@ def get_video_replacement_qwen3vl(item_idx: int):
             ),
         ]
 
+    @staticmethod
+    def get_video_repl(
```
Comment (2ez4bz) [inline vllm/model_executor/models/qwen3_vl.py:1168]: Similar to #27107, which added support for EVS for Nemotron Nano V2 VL, we need to switch the prompt replacement logic to take interleaved tokens into account. This helper is used:

1. in preprocessing when doing prompt replacement, where the `sum(tokens_per_frame)` needs to be consistent with what we will have at runtime when merging the embeddings.
2. and when computing the final video embeddings post pruning, when the actual `tokens_per_frame` will be known.

File: vllm/model_executor/models/qwen3_vl.py [hunk 4]
```diff
@@ -1471,88 +1591,304 @@ def _postprocess_image_embeds_evs(
...
+        # Expand positions to match the full sequence length
+        # (includes both video tokens and indicator tokens)
+        # Shape: [full_length, 4] where positions are filled for video tokens
+        # and zeros for indicator tokens, EXCEPT channel 3 (max width) which
+        # must be filled for all tokens so recompute_mrope_positions can extract it
+        expanded_positions = torch.zeros(
+            seq_len,
+            5,  # 5 channels: [t_index, h_index, w_index, max_width, is_video]
+            device=device,
+            dtype=torch.long,
```
Comment (2ez4bz) [inline vllm/model_executor/models/qwen3_vl.py:1816]: Note: in a prior version of EVS for Qwen3 VL that I was working on, I was borrowing from the qwen2.5 VL approach and "smuggling" these as extra channels in the embeddings tensor. However, this is inherently loss, as embeddings are bf16 / fp16, or even worse, fp8.

File: vllm/model_executor/models/qwen3_vl.py [hunk 5]
```diff
@@ -1573,66 +1909,77 @@ def _parse_and_validate_multimodal_inputs(self, **kwargs: object) -> dict:
...
-        self, input_tokens: list[int], mm_features: list[MultiModalFeatureSpec]
-    ) -> Iterator[tuple[int, int, int]]:
-        """
-        Iterate over multimodal features and yield grid information.
-
-        For videos with EVS (Efficient Video Sampling) enabled, this function
-        computes the offset based on the pruned token count rather than relying
-        on input_tokens.index(), which would fail when tokens are pruned.
+    @staticmethod
+    def _iter_mm_grid_hw(
```
Comment (2ez4bz) [inline vllm/model_executor/models/qwen3_vl.py:1881]: This was broken out into a static method for easier testing; I could not for the life of me figure out how to instantiate a `Qwen3VLForConditionalGeneration` for direct testing.

File: vllm/model_executor/models/qwen3_vl.py [hunk 6]
```diff
@@ -1737,13 +2084,101 @@ def _get_actual_frame_token_counts(
...
+        mm_features: list[MultiModalFeatureSpec],
+    ) -> tuple[torch.Tensor, int]:
+        return self._get_mrope_input_positions(
+            input_tokens=input_tokens,
+            mm_features=mm_features,
+            config=self.config,
+        )
+
+    @staticmethod
+    def _get_mrope_input_positions(
```
Comment (2ez4bz) [inline vllm/model_executor/models/qwen3_vl.py:2067]: Ditto re testing.

File: vllm/model_executor/models/qwen3_vl.py [hunk 7]
```diff
@@ -1763,10 +2201,28 @@ def recompute_mrope_positions(
...
+            multimodal_positions=multimodal_positions,
+            mrope_positions=mrope_positions,
+            num_computed_tokens=num_computed_tokens,
+            image_token_id=self.config.image_token_id,
+            video_token_id=self.config.video_token_id,
+            vision_start_token_id=self.config.vision_start_token_id,
+        )
 
+    @staticmethod
+    def _recompute_mrope_positions(
```
Comment (2ez4bz) [inline vllm/model_executor/models/qwen3_vl.py:2180]: Ditto re testing.

File: vllm/v1/worker/gpu_model_runner.py [hunk 8]
```diff
@@ -2425,23 +2423,65 @@ def _execute_mm_encoder(
...
                     curr_group_outputs = model.embed_multimodal(**mm_kwargs_group)
+                    (
+                        curr_group_outputs,
+                        postprocess,
+                        curr_group_positions,
+                    ) = self._unwrap_mm_outputs(  # type: ignore
+                        curr_group_outputs
+                    )
+                    if postprocess is not None:
+                        curr_group_outputs = postprocess(curr_group_outputs)
```
Comment (2ez4bz) [inline vllm/v1/worker/gpu_model_runner.py]: This is the GPU-memory related hack mentioned in the `interfaces.py` comment. The `postprocess` callback returned by Qwen3-VL results in additional memory being taken, which, without this workaround, leads the profile step thinking less tokens can fit in KV cache.

If anybody has suggestions for a better fix (which ideally, would just be encapsulated within the `Qwen3VLForConditionalGeneration` code that computes the final embeddings), I am happy to consider them :)

File: pyproject.toml [hunk 9]
```diff
@@ -148,7 +148,7 @@ locale = "en"
 extend-ignore-identifiers-re = ["NVML_*", ".*Unc.*", ".*_thw",
     ".*UE8M0.*", ".*[UE4M3|ue4m3].*", ".*eles.*",
      ".*[Tt]h[rR].*"]
-extend-ignore-words-re = []
+extend-ignore-words-re = ["thw"]
```
Comment (2ez4bz) [inline pyproject.toml]: Added because of using `thw` for mrope related unit tests.

File: vllm/model_executor/models/interfaces.py (hunk 2 above)
Comment (2ez4bz) [reply vllm/model_executor/models/interfaces.py]: No longer relevant.

File: vllm/v1/worker/gpu_model_runner.py (hunk 8 above)
Comment (2ez4bz) [reply vllm/v1/worker/gpu_model_runner.py]: No longer relevant.

File: pyproject.toml (hunk 9 above)
Comment (DarkLight1337) [reply pyproject.toml]: Can you rename the variable to `grid_thw` to avoid having to change the config?

File: tests/model_executor/test_qwen3_vl_mrope.py [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+    expected_mrope_masked = expected_mrope[:, whole_sequence_retention_mask]
+
+    # Initialize computed_mrope with sequential positions for all prefix tokens
+    computed_mrope = torch.empty((3, len(input_tokens_pruned)), dtype=torch.long)
+    computed_mrope[:, 0 : len(prefix_tokens)] = expected_mrope[
+        :, 0 : len(prefix_tokens)
+    ]
+
+    # Paranoia check that computed_mrope is wrong.
+    assert torch.equal(computed_mrope, expected_mrope_masked) is False
```
Comment (DarkLight1337) [inline tests/model_executor/test_qwen3_vl_mrope.py]:
```suggestion
    assert not torch.equal(computed_mrope, expected_mrope_masked)
```

File: vllm/model_executor/models/qwen3_vl_moe.py [hunk 11]
```diff
@@ -437,6 +437,7 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
         multimodal_config = vllm_config.model_config.multimodal_config
 
         self.config = config
+        self._model_config = vllm_config.model_config
```
Comment (DarkLight1337) [inline vllm/model_executor/models/qwen3_vl_moe.py]: We can simply create the tokenizer in `__init__` method

File: vllm/model_executor/models/qwen3_vl.py [hunk 12]
```diff
@@ -1502,62 +1616,214 @@ def _postprocess_video_embeds_evs(
...
+            timestamps=timestamps,
+            vision_start_token_id=self.config.vision_start_token_id,
+            vision_end_token_id=self.config.vision_end_token_id,
+            video_token_id=self.config.video_token_id,
+            select_token_id=self.is_multimodal_pruning_enabled,
+        )
+
+        repl_token_ids = torch.tensor(video_repl.full, device=device)
+        embed_token_id = _cached_tensor(self.config.video_token_id, device=device)
+        is_video_embed = torch.isin(repl_token_ids, embed_token_id)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/qwen3_vl.py:1740]: To avoid blocking model forward, I think it's better to calculate this inside the processor's `_call_hf_processor`, and add a corresponding field to `_get_mm_fields_config`.

File: tests/model_executor/test_qwen3_vl_mrope.py (hunk 10 above)
Comment (2ez4bz) [reply tests/model_executor/test_qwen3_vl_mrope.py]: Done.

File: vllm/model_executor/models/qwen3_vl.py (hunk 12 above)
Comment (2ez4bz) [reply vllm/model_executor/models/qwen3_vl.py:1740]: Unless I'm mistaken, I don't think that would work, unfortunately.

`is_video_embed` (and all other tensors here based off of `video_repl.full`) depends on the exact token sequence, which is not known at preprocessing time. Unlike Qwen2.5 VL, where there were no timestamp tokens, we cannot assume there is a single contiguous chunk of vision placeholder tokens here. So, even though the token count remains the same between preprocessing and this part of the code, the actual sequence of tokens does not.

Let me know if you meant something else / saw a way to make that work, though, as it would greatly simplify the code.

File: vllm/model_executor/models/qwen3_vl_moe.py (hunk 11 above)
Comment (2ez4bz) [reply vllm/model_executor/models/qwen3_vl_moe.py]: Removed in favor of `self._tokenizer = cached_tokenizer_from_config(vllm_config.model_config)`.

File: pyproject.toml (hunk 9 above)
Comment (2ez4bz) [reply pyproject.toml]: Done.

File: vllm/model_executor/models/qwen3_vl.py [hunk 13]
```diff
@@ -1471,88 +1591,304 @@ def _postprocess_image_embeds_evs(
...
+        )
+        full_is_video_embed = unpruned_token_ids_tensor == embed_token_id
+        expanded_positions[is_video_embed, :3] = original_mrope[full_is_video_embed][
+            retention_mask
+        ]
+        expanded_positions[~is_video_embed, :3] = original_mrope[~full_is_video_embed]
+        expanded_positions[..., 3] = w // merge_size
+        expanded_positions[..., 4] = is_video_embed
+
+        return expanded_positions
```
[the remark this answers is not in the document]
Comment (2ez4bz) [reply vllm/model_executor/models/qwen3_vl.py:1859]: No longer relevant.

File: vllm/model_executor/models/qwen3_vl.py (hunk 12 above)
Comment (DarkLight1337) [reply vllm/model_executor/models/qwen3_vl.py:1740]: Perhaps you could call a reduced version of `_get_mm_prompt_updates` inside `_call_hf_processor`? You should already have the necessary data after HF processor is called.

Comment (DarkLight1337) [reply vllm/model_executor/models/qwen3_vl.py:1740]: Discussed offline, there is no other choice

**Review by ywang96 — APPROVED** [review]
